### PR TITLE
Add preact support

### PR DESCRIPTION
**Pull request by tkh44**
**What**:

**Why**:

**How**:

**Checklist**:

- [ ] Documentation
- [x] Tests
- [x] Code complete

### CODE REVIEW

File: src/preact.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export default function (tag, cls, vars = [], content) {
+  if (!tag) {
+    throw new Error(
+      'You are trying to create a styled element with an undefined component.\nYou may have forgotten to import it.'
+    )
+  }
+
+  function Styled (props) {
+    const getValue = v => (v && typeof v === 'function' ? v.cls || v(props) : v)
```
Comment (lukeed) [inline src/preact.js]: I'd hoist this, redefined per `Styled` tag.

File: src/preact.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      cls,
+      vars.map(getValue),
+      content
+    )
+
+    return h(
+      tag,
+      Object.assign({}, props, {
+        ref: props.innerRef,
+        class: props.class
```
Comment (lukeed) [inline src/preact.js]: Should use `classname` key. Also can be this:

```js
classname: `${className} ${props.class || ''}`
```

File: src/preact.js (hunk 1 above)
Comment (tkh44) [inline src/preact.js]: Can't because of props.

Comment (lukeed) [inline src/preact.js]: Can still hoist and pass, but looking at it again, is just a style thing.

**Comment by ooade** [comment]
Dropping this?

**Comment by tkh44** [comment]
It's too much to maintain. I wouldn't mind someone creating an external version of `styled` for preact.

**Comment by ooade** [comment]
Okay. Will try taking this when i'm free.

**Comment by Connorelsea** [comment]
Sad to see this closed. What would the best alternative be for use with Preact?

**Comment by mitchellhamilton** [comment]
@Connorelsea emotion should work fine with preact-compat or _maybe_ even just aliased to preact.

**Comment by developit** [comment]
We can add some guides on how to do the latter. I don't think this even uses any part of compat, it's more just the package name.

**Comment by mitchellhamilton** [comment]
The package we use for theming uses `React.Children.only` so I think we might need compat. AFAIK emotion should work fine without compat if you don't use theming

https://github.com/iamstarkov/theming/blob/master/src/create-theme-provider.js#L96

**Comment by Connorelsea** [comment]
Seems to have mismatched class names when using the `emotion/babel` plugin with react-cli, unless I am doing something irrelevant incorrectly. See the below picture for example:

image
image

**Comment by mitchellhamilton** [comment]
I'm guessing you're using `preact-cli`, IIRC in the components folder css files are imported as css modules

**Comment by developit** [comment]
yup yup. Something like this would work:

```js
export default (config, env, helpers) => {
  let babelConfig - helpers.getLoadersByName(config, 'babel-loader')[0].rule.options;
  babelConfig.plugins.push('emotion/babel');

  // turn off CSS Modules:
  helpers.getRulesByMatchingFile('src/components/x.css').rule.test = () => false;
  delete helpers.getRulesByMatchingFile('x.css').rule.exclude;
};







```

**Comment by tkh44** [comment]
I would recommend putting `emotion/babel` first in your list. It needs to come before a couple of babel plugins.

**Comment by developit** [comment]
If the above works, it can be published to npm as `preact-cli-plugin-emotion`. The next version of the CLI will support plugins via package.json instead of requiring they be invoked from config.

**Comment by Connorelsea** [comment]
Nice, sounds good. Also, enabling inline-mode works as well, dunno what the benefits/cons to this are though.

```javascript
  let { rule } = helpers.getLoadersByName(config, "babel-loader")[0]
  let babelConfig = rule.options

  babelConfig.plugins.push(["emotion/babel", { inline: true }])
```

**Comment by tkh44** [comment]
I would run extract mode unless you need ie11 support. I'm not sure you can extract the critical css needed to run inline mode in a server rendered env.

**Pull request closed without merging**